### PR TITLE
Ensure charging station lastSeen is updated only if the station exists

### PR DESCRIPTION
**Pull request by jerome-benoit**
in DB, is not deleted and not a roaming one.

Create indexes to speed up the query (also filters on logs collection, a lot ...)

Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

### CODE REVIEW

File: src/server/ocpp/services/OCPPService.ts [hunk 1]
```diff
@@ -87,7 +87,7 @@ export default class OCPPService {
       bootNotification.lastSeen = bootNotification.lastReboot;
       bootNotification.timestamp = bootNotification.lastReboot;
       // Get the charging station
-      let chargingStation = await ChargingStationStorage.getChargingStation(headers.tenantID, headers.chargeBoxIdentity);
+      let chargingStation: ChargingStation = await ChargingStationStorage.getChargingStation(headers.tenantID, headers.chargeBoxIdentity);
```
Comment (LucasBrazi06) [inline src/server/ocpp/services/OCPPService.ts:90]: No need to provide the type here, already provided by the method signature

File: src/storage/mongodb/MongoDBStorage.ts [hunk 2]
```diff
@@ -258,6 +260,11 @@ export default class MongoDBStorage {
     // Logs
     await this.handleIndexesInCollection(Constants.DEFAULT_TENANT, 'logs', [
       { fields: { timestamp: 1 } },
+      { fields: { type: 1 } },
```
Comment (LucasBrazi06) [inline src/storage/mongodb/MongoDBStorage.ts:263]: Index 'type' is already handled in index { fields: { type: 1, timestamp: 1 } }
Same for 'action', 'level', source and host.

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]
Check my comments

**Comment by LucasBrazi06** [comment]
Another PR with 24 files changed instead of 1 related to the title of this PR.
You still continue to not following the rules in place.
I already issued several warnings to you and you still continue to do what you want.
I don't know what to do make you listen in the end!
So I'll close this PR and you will create several PRs with clear purpose following the rules I put in place.
I hope you'll finally start learning your lessons.

File: src/storage/mongodb/MongoDBStorage.ts (hunk 2 above)
Comment (jerome-benoit) [reply src/storage/mongodb/MongoDBStorage.ts:263]: Indexes are per field https://docs.mongodb.com/manual/core/index-single/ and compound indexes are only specific all fields together with an importance to the order: https://docs.mongodb.com/manual/core/index-compound

Comment (LucasBrazi06) [reply src/storage/mongodb/MongoDBStorage.ts:263]: Not at all!
image

**Comment by jerome-benoit** [comment]
> Another PR with 24 files changed instead of 1 related to the title of this PR.
> You still continue to not following the rules in place.
> 

The code flow implied by the rules is not really efficient because does not follow the git design by confusing the different levels of code change tracking, reviewing and merging.
Efficient code flow examples are available in various main FOSS project, the main ones being used on git development or Linux kernel development.
The hope it will be noticed on his own at some point have not been fulfilled, unfortunately.

> I already issued several warnings to you and you still continue to do what you want.
> I don't know what to do make you listen in the end!
> So I'll close this PR and you will create several PRs with clear purpose following the rules I put in place.
> I hope you'll finally start learning your lessons.
> 


Do you really think that a more than 30 years FOSS developers need explanations on how to make their code changes acceptable and reviewable?
A sensible approach would have been to rely on their knowledge to improve the existing code flow with git with an atomic change review at the commit level.
In the team, there’s now experienced developers that can suggest improvements on git usage or rules that are counter-productive and inhibit developers will at making the code better easily, relying on their knowledge could be a wise idea.

I do not have time currently to do an unproductive split by PR of the atomic and thematic changes at the commit level in that branch. Fixes will wait.

--
Jérôme Benoit aka fraggle
Piment Noir - https://piment-noir.org
OpenPGP Key ID : 27B535D3
Key fingerprint : B799 BBF6 8EC8 911B B8D7 CDBC C3B1 92C6 27B5 35D3

**Pull request closed without merging**